### ISSUE TITLE
[Feature Request] Invitation System

### ISSUE THREAD

**Issue by tinloaf** [syncthing/syncthing#2300]
Hi,

I would love to work towards replacing Dropbox with Syncthing, only that trying to walk people through "add my device, then I add you back (yes you have to copy that code and send it to me!), then you can add the folder (no, it has to be spelled exactly like this! Dont change anything else! No you are not a master!) and as soon as I added you to the folder, everything works!" mostly results in "Can't we just use Dropbox?"

Here's how "Invitations" would come in, in three stages:

**Stage One**: Syncthing is able to open a special kind of XML files (and associated with those files on the operating system). These XML files contain two sections:
- One section containing informations about devices to pair with (Key/ID, name, discovery server used?)
- One section containing informations about folders to be added (Folder ID, devices that this folder exists on (referencing the pre vious section))

If Syncthing is told to open one of these files, a dialogue should pop up (I know, that's hard to do with a web interface) asking the user for each folder in the XML file what local folder to use (cautioning that all contents of that folder will be shared with the devices listed in the device-section). *All other settings should be hidden*, perhaps collapsed and expandable via "Extended Settings" or whatever. After that, the own ID should be displayed with a message "please send this ID to whomever you got that file from".

This would make the use case of "me adding someone to my shared folder" a two-step process on the receiving side (open file, confirm, send displayed ID).

 **Stage Two**: On top of that, every device entry in the XML file can optionally contain a one-time (secret) key used for pairing: The client opening that file (call it device A) signs his own public key (aka "A's ID") with that one-time key and sends it to the device that the one-time key is associated with (device B). B then adds A and shares preconfigured folders with it (preferably those listed in the XML file) with A, without requiring any user intervention. This would make invitation a one-step process. On the other hand, anyone able to intercept the XML file will be able to gain access to the share. But, realistically the XML file as well as the returning "My devices ID is XYZ" will both be sent via the same medium, so someone able to intercept the XML file is probably also able to tamper with the device ID sent back. However, including this "authorization token" should be strictly optional. If no such token is present, the ID has to be added manually on device B.

**Stage Three**: Build a nice UI for creating these XML files in Syncthing. This would be more or less necessary after Stage Two anyway, since the authorization tokens (and what folders they authorize access to) must be somehow stored in the Syncthing instance issuing these tokens.

If something like this was implemented, I think handling sharing files with Syncthing would be just as easy as with Dropbox (or whatever). I do realize this is a rather huge implementation request (and I can't write a word of Go, or I would possibly even help ;)), but I'd like to discuss this.

**Comment by AudriusButkevicius** [comment]
We already have prompts when devices connect and suggest to share a folder (so no name typing) so I don't see how this is different apart from locking down the options, and having to toss around files.

**Comment by tinloaf** [comment]
Well, for the pairing prompt to show up, one side has to manually type in the other side's ID. 

- If I tell my friend to enter my ID manually, he will still have to send back his ID, since I will hardly just accept any random ID that wants to pair with me. So, no gain there. 
- If I manually enter his ID, I still have to ask him to send me his ID (step 1), then accept my pairing, verifying my key in the process (Step 2). 

After that, I can then share the folder with him, but to accept that, my friend still has to actively open the web interface again and accept the folder (Step 3).

In my experience, this often is an "asynchronous" process, i.e. I send an email, wait a couple of hours for a response, then take the next step. Thus, going through all 3 steps might take quite a while, enough time for people to lament how much easier this is with Dropbox (which it is! Just input the other sides' email address, and everything is done).

Don't you think that just double-clicking an XML, confirming a dialogue and seeing how synchronization starts a minute after this will make for an improved user experience? Don't get me wrong here, I don't mind inputting a few IDs now and then, but convincing lazy people to switch their preferred solution to something else is usually a really hard task if that solution comes with more effort than the current solution...

**Comment by tinloaf** [comment]
Also, regarding the tossing around files thing: If Syncthing had a possibility to create these files from its user interface, why not just serve them from the REST API afterwards? This way, the user creating the invitation would get a link with some token, sends that link via email. The other side then clicks the link, it opens in a browser, triggering a download of mimetype 'text/syncthing-whatever', which syncthing is configured to handle. No file-tossing anywhere.

**Comment by marrie** [comment]
I like the idea, but it seems a bit confusing. Are you saying to do the following
* generate the xml file  on your end
* email the link to the person or persons of interest
* they except the xml file and click it and it some how opens up in syncthing? 

That could work but might I 
add maybe a  way to generate a list of folders in a  text file that you could send out and they tell you what folders to add to the xml? or am I just confusing things even more.

**Comment by marrie** [comment]
I should add I would also love this feature as I have one user that just doesn't get syncthing, won't use dropbox, uses sendspace, and wants me to email a 29 gig zip file.  Yeah I think the xml might be easier for this person. Open syncthing, run this xml and follow prompts. then sit back and wait for the files you wanted  to get to you.

**Comment by kozec** [comment]
There was discussion about very similar idea, but with URL instead of XML file. In plain theory, there should be possible to use `registerProtocolHandler` js function to register syncthing:// scheme and then pass everything needed in one, long url sent, for example, through im...

**Comment by marrie** [comment]
Would this be accessible from the web? I mean if someone is in Florida and I give them this link after  they install syncthing woudl it then connect to my computer and they can except the folders and my device accepts them etc? yeah the link sounds like a good idea.

**Comment by kozec** [comment]
@marrie Well, it should be possible, theoretically. WebUI can register 'web+syncthing://' protocol (chrome refuses to register it without web+ prefix) and parse whatever is in address when called.

I've just tested small proof-of-concept where clicking on `web+syncthing://test` link opens WebUI with prefilled Add Device dialog. It works even with xdg-open, so it should work with IM and similar apps, as long as they can recognize it as uri.

[Here, if anyone is interested](https://github.com/kozec/syncthing/commit/d1bda0d807cbf391d512d1cf48e706e467574f34), but be warned, this was first time I touched AngularJs.

But, apart from security concerns, there is one problem with this approach. Uri with device id, folder id and names will be rather long and ugly.

**Comment by marrie** [comment]
Yeah that's true, unless it can generate a file that is then downloaded to the user's computer, this way it's not as ugly. then the user just double clickes the file and it opens up in the web UI. I've seen ugly links that are so long I would not want to memorise them, but the end user does not in my opinion need to know the link is that long and that ugly..

**Comment by AudriusButkevicius** [comment]
Why do you think that registering a file handler is easy? It's much harder than the example by kozec, especially when you have to deal with every breed of every OS who use different window managers etc. Also, sending a file is extra steps, having to attach it somewhere, download it on the receiving end, find downloads folder, then for example pasting a link in an irc message. 

Also, its more scary to run some random file, than it is to click on a link, atleast to me. You can wrap the URL in a <a> tag, and hide the uglyness. 

Regardless, I don't understand how kozec example works to decide if its feasible. I guess it only works while the tab with the handler installed is open.

**Comment by kozec** [comment]
> Regardless, I don't understand how kozec example works to decide if its feasible. I guess it only works while the tab with the handler installed is open.

Nope, it always opens a new tab, even if WebUI is already open somewhere. `registerProtocolHandler` simply sets address, that will be open when uri with that scheme is requested.

[It is documented here](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler)

**Comment by marrie** [comment]
I'm someone who finds clicking on files easier locally, but yeah give the  ugly link a try and maybe we can get something working for invites and stuff.

Take care.

**Comment by AudriusButkevicius** [comment]
This breaks when you have multiple instances of syncthing I guess.

**Comment by marrie** [comment]
Ah I only have one instence running on my mac.  so for me tha tmight work. It's worth a try anyway.

**Comment by AudriusButkevicius** [comment]
Feel free to make a pull request.

**Comment by marrie** [comment]
Agreed.  it sounds like it could work. I'll be keeping  an eye on this.

Blessings and happy Wednesday

**Comment by haikion** [comment]
Something similar to Bittorrent Syncs "read only key" would be useful. This could be integrated into protocol handler in order to have "syncthing://key" links on a web page. Clicking link would open Syncthing dialog in which user could choose download directory. No additional input from server side should be required.

**Comment by AudriusButkevicius** [comment]
I think Bittorrent Syncs has moved away from this approach due to it's flaws of the key leaking once, and it being impossible to revoke it.

**Comment by haikion** [comment]
BitTorrent Sync still uses this method. By default it adds an expiration duration to the key but it is possible to create eternal keys. I don't know how BitTorrent Sync handles manual key revocation or if it even allows it.
In Syncthing implementation multiple public keys could be generated from a single private key. This means one public key could be revoked without revoking others.

**Comment by AudriusButkevicius** [comment]
There is no way do have multiple public keys for a private key 
Also, public keys are not 32 characters long to be easy to throw around.

It's not clear how BTSync implements it to be able to mimic it.

**Comment by haikion** [comment]
> There is no way do have multiple public keys for a private key 

It appears I had some incorrect assumptions regarding this, sorry.

**Comment by kalsan** [comment]
I love the ideas in this thread and I'd like to revive it with a fresh idea. My motivation is the following scenario: my friend wants a huuuuge set of laaaarge files from me but she's on the other side of the planet (actually "city" would be more appropriate, but we're lazy folks lacking the ability to carry HDDs, so let both be eqivalent). Right now, we do it like this: she sends me her ID, I add it to my server and share my stuff folder and she eventually gets an invitation for both (sometimes with a significant delay in between). With the invitation system, I could tell my syncthing to generate an invite to the folder. It gives me either a link (as discussed above), a file (discussed as well) or, new idea (I think), a magic string containing my ST ID, maybe a pointer to the folder and, most importantly, a nonce. She now goes to her WebGUI and simply pastes the magic string, pairing done.

So what's new is the nonce. My ST node keeps a set of all nonces that it ever sent and that are still valid. It is imaginable that nonces can have properties that can be set when creating the invite, for instance:

- expiry: if a nonce was not used until yyyy-mm-dd it will automatically expire and my ST will forget it
- max usages: this defaults to 1, meaning that a single device may pair using this nonce. However I can set it to 5 (the nonce expires when the 5th ST ID has successfully paired), or to infinite if I want to make a folder public
- device filter: I can tell my ST that the nounce is only valid for a given set of other ST IDs

This is a much simpler approach than generating key pairs. All that is needed is a map nonce -> permissions in Syncthing. However, the pairing process would need to be updated, i.e. the ST protocol would need to support sending the encrypted nonce to a host when requesting to pair.

This is just a quick idea. Let me know what you think folks.

**Comment by AudriusButkevicius** [comment]
You don't need to wait for invitations or anything, she can just add your device and your folder on the other end and it should just work, that is all the prompt does, adds the device with that id and folder with that id, so it's enough for you just to email device id and folder id.

**Comment by kalsan** [comment]
You are right, it can be done in one step and probably even be scripted to be added to the ST config / via API. However, things such as expiry, max usages and device wildcard would not be possible.

**Comment by ThiloteE** [comment]
Bump. Sorry, saw too late that this thread existed.

From a marketing perspective, this is probably one of the most important features syncthing could have to reach the mass-market and non-tech-savvy folks: 

**A link** that can be sent to people for them to open and connect to a folder or a file in the browser.

- I dare to reason that the huge success of programmes like Discord, Zoom or Dropbox, was at least partly to **easy accessibility**.
- Clicking a link to join a meeting (joining a folder in this case) and some follow up clicks in the browser is a VERY easy way to access a service.
-  In particular, it omits having to install Syncthing and it omits a lot of the configuration, which, for many, is a technical challenge and in this case necessitates to browse the Syncthing website to find out how to do so. My parents would not know how to do it, just saying. More importantly, they would not WANT to do it.
- It would allow a tech-savvy person to do the initial setup without having to force every short-term acquaintance to install Syncthing.
- it would help to simplify a situation, in which the sender does not have access to the receiving devices.
 
 Sorry, i wouldn't know how to program this though. 😅
